src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -0,0 +1,223 @@
+        } else {
+            throw new UnsupportedOperationException("dynamic table is not implemented");
+        }
+    }
+
+    private void decodeLiteral(ByteBuf in, Sink sink) throws Http3Exception {
+        final CharSequence name = decodePrefixedStringLiteral(in, (byte) 0x8, 3);
+        final CharSequence value = decodePrefixedStringLiteral(in);
+        sink.appendToHeaderList(name, value);
+    }
+
+    private CharSequence decodePrefixedStringLiteral(ByteBuf in) throws Http3Exception {
+        return decodePrefixedStringLiteral(in, (byte) 0x80, 7);
+    }
+
+    private CharSequence decodePrefixedStringLiteral(ByteBuf in, byte mask, int prefix) throws Http3Exception {
+        final boolean huffmanEncoded = (in.getByte(in.readerIndex()) & mask) == mask;
+        final int length = (int) decodePrefixedInteger(in, prefix);
+        return decodeStringLiteral(in, length, huffmanEncoded);
+    }
+
+    // TODO: not sure if integer length is enough
+    private CharSequence decodeStringLiteral(ByteBuf in, int length, boolean huffmanEncoded) throws Http3Exception {
+        if (huffmanEncoded) {
+            return huffmanDecoder.decode(in, length);
+        }
+        byte[] buf = new byte[length];
+        in.readBytes(buf);
+        return new AsciiString(buf, false);
+    }
+
+    private CharSequence getIndexedName(int index) throws Http3Exception {
+        if (index <= QpackStaticTable.length) {
+            final QpackHeaderField field = QpackStaticTable.getField(index);
+            return field.name;
+        }
+        throw NAME_ILLEGAL_INDEX_VALUE;
+    }
+
+    private QpackHeaderField getIndexedHeader(int index) throws Http3Exception {
+        if (index <= QpackStaticTable.length) {
+            return QpackStaticTable.getField(index);
+        }
+        throw HEADER_ILLEGAL_INDEX_VALUE;
+    }
+
+    // private static HeaderType validate(CharSequence name, HeaderType previousHeaderType) throws Http3Exception {
+    //     // TODO: implement
+    //     return previousHeaderType;
+    // }
+
+    private interface Sink {
+        void appendToHeaderList(CharSequence name, CharSequence value);
+        void finish() throws Http3Exception;
+    }
+
+    private static final class Http3HeadersSink implements Sink {
+        private final Http3Headers headers;
+        private final long maxHeaderListSize;
+        private final boolean validate;
+        private long headersLength;
+        private boolean exceededMaxLength;
+        private Http3Exception validationException;
+
+        Http3HeadersSink(Http3Headers headers, long maxHeaderListSize, boolean validate) {
+            this.headers = headers;
+            this.maxHeaderListSize = maxHeaderListSize;
+            this.validate = validate;
+        }
+
+        @Override
+        public void finish() throws Http3Exception {
+            if (exceededMaxLength) {
+                // TODO: check would require information from peer?
```

Comment:
You mean we would need to have informations related to the settings that the peer did sent ?

Reply:
Yes, I suppose so. Was going to check spec more closely on this

Reply:
yeah I think thats the case... I will plug this in when I implement http3 handling

Reply:
We can extend `decodeHeaders` function with additional argument for max size. This way decoder won't  depend on where and how those settings are stored/synced

Reply:
Okay, so this configuration is actually only local. The configuration that's exchanged with peer is max table size not the size of a single header fields block. Keeping the logic same as in HPACK